File: lv1-library-loan/csharp/LibraryLoan.Application/UseCases/BorrowBookUseCase.cs
using LibraryLoan.Application.Books;
using LibraryLoan.Application.Members;
using LibraryLoan.Domain.Books;
using LibraryLoan.Domain.Members;
using LibraryLoan.Results;

namespace LibraryLoan.Application.UseCases;

/// <summary>ユースケース: 会員が本を借りる。</summary>
public sealed class BorrowBookUseCase
{
    private readonly IMemberRepository _memberRepository;
    private readonly IBookRepository _bookRepository;

    public BorrowBookUseCase(IMemberRepository memberRepository, IBookRepository bookRepository)
    {
        _memberRepository = memberRepository;
        _bookRepository = bookRepository;
    }

    public Result<LoanRecordId> Handle(MemberId memberId, BookId bookId, DateOnly loanDate)
    {
        return FindMember(memberId).Bind(member =>
            FindBook(bookId).Bind(_ =>
                member.Borrow(LoanRecordId.NewId(), bookId, loanDate).Map(loanRecord =>
                {
                    _memberRepository.Save(member);
                    return loanRecord.Id;
                })));
    }

    private Result<Member> FindMember(MemberId memberId) =>
        _memberRepository.Find(memberId) is { } member
            ? member
            : Result<Member>.Failure(MemberRepositoryErrors.NotFound(memberId));

    private Result<Book> FindBook(BookId bookId) =>
        _bookRepository.Find(bookId) is { } book
            ? book
            : Result<Book>.Failure(BookRepositoryErrors.NotFound(bookId));
}

File: lv1-library-loan/typescript/src/application/use-cases/borrow-book-use-case.ts
import type { BookRepository } from "../books/book-repository.js";
import { BookRepositoryErrors } from "../books/book-repository-errors.js";
import type { Book } from "../../domain/books/book.js";
import type { BookId } from "../../domain/books/book-id.js";
import type { Member } from "../../domain/members/member.js";
import type { MemberId } from "../../domain/members/member-id.js";
import { LoanRecordId } from "../../domain/members/loan-record-id.js";
import type { LocalDate } from "../../domain/shared/local-date.js";
import { Result } from "../../results/result.js";
import type { MemberRepository } from "../members/member-repository.js";
import { MemberRepositoryErrors } from "../members/member-repository-errors.js";

/** ユースケース: 会員が本を借りる。 */
export class BorrowBookUseCase {
  constructor(
    private readonly memberRepository: MemberRepository,
    private readonly bookRepository: BookRepository,
  ) {}

  handle(memberId: MemberId, bookId: BookId, loanDate: LocalDate): Result<LoanRecordId> {
    return this.findMember(memberId).bind((member) =>
      this.findBook(bookId).bind(() =>
        member.borrow(LoanRecordId.newId(), bookId, loanDate).map((loanRecord) => {
          this.memberRepository.save(member);
          return loanRecord.id;
        }),
      ),
    );
  }

  private findMember(memberId: MemberId): Result<Member> {
    const member = this.memberRepository.find(memberId);
    return member !== undefined ? Result.success(member) : Result.failure(MemberRepositoryErrors.notFound(memberId));
  }

  private findBook(bookId: BookId): Result<Book> {
    const book = this.bookRepository.find(bookId);
    return book !== undefined ? Result.success(book) : Result.failure(BookRepositoryErrors.notFound(bookId));
  }
}
